### PR TITLE
feat: add SSRF protection for df.http()

### PR DESCRIPTION
**Pull request by pinodeca**
## Summary

Add compile-time dataplane SSRF protection that blocks HTTP requests to private/reserved IP ranges from the pg_durable background worker. This prevents malicious users from using `df.http()` to probe internal network services, cloud metadata endpoints (169.254.169.254), or localhost services.

Addresses **threat T8** (CRITICAL) in spec-security-model.md.

### New: `src/ssrf.rs`
- IP blocklist covering RFC 1918, loopback, link-local, ULA, and unspecified ranges
- IPv4-mapped IPv6 handling (`::ffff:A.B.C.D` → extract and check IPv4)
- URL scheme validation (HTTP/HTTPS only)
- Pre-flight IP-literal check (`validate_url_host`) — reqwest skips DNS resolution for IP literal URLs, so the resolver alone is insufficient
- Custom DNS resolver (`SsrfSafeResolver`) that checks resolved IPs before TCP connect — prevents DNS rebinding attacks
- Comprehensive unit tests (25 tests covering all ranges, edge cases, IPv6 variants, IP literals)

### Modified: `src/activities/execute_http.rs`
- Scheme validation before any network I/O
- IP-literal validation before reqwest sees the URL
- SSRF-safe DNS resolver injected into reqwest client
- Structured `HTTP BLOCKED (scheme)` and `HTTP BLOCKED (ip)` audit logs with `submitted_by` and `login_role`

### Modified: `src/orchestrations/execute_function_graph.rs`
- Thread `submitted_by` and `login_role` from `FunctionNode` into `HttpConfig` JSON

### Modified: `src/types.rs`
- Add `submitted_by` and `login_role` fields to `HttpConfig`

### Modified: `Cargo.toml`
- SSRF protection is on by default (no feature flag needed)
- New `no-ssrf-protection` Cargo feature to opt out for local dev/testing

### Modified: `.github/workflows/ci.yml`
- Clippy runs in both modes: default (SSRF on) and `no-ssrf-protection`

### Modified: `docs/spec-ssrf-protection.md`
- Status updated to **Completed**
- Feature flag docs updated to reflect `no-ssrf-protection` inversion

### Modified: `docs/spec-security-model.md`
- T8 status updated to **Implemented**
- Trimmed duplicated SSRF implementation detail, now references spec-ssrf-protection.md

### New: `tests/e2e/sql/36_ssrf_protection.sql`
- E2E test with 4 cases: block metadata endpoint (169.254.169.254), block localhost (127.0.0.1), block `file://` scheme, allow legitimate public HTTPS

## Design

**Two-layer model:**
1. **Layer 1 (this PR):** Compile-time dataplane protection — hardcoded IP blocklist, no GUC/superuser override possible
2. **Layer 2 (future):** Customer-level controls — URL allowlists, REVOKE EXECUTE, rate limiting

**Key decisions:**
- SSRF protection is on by default with no feature flag needed; `no-ssrf-protection` is an opt-in escape hatch
- `--no-default-features` builds (e.g., CI) retain SSRF protection
- Both IP literals and DNS-resolved hostnames are checked (two enforcement points)
- DNS rebinding prevented by checking IPs inline in the resolver (same address used for check and connect)
- Error messages do not leak resolved IPs
- CI verifies compilation in both modes

### CODE REVIEW

File: docs/spec-ssrf-protection.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+
+```toml
+[features]
+default = ["pg17"]
+no-ssrf-protection = []
+```
+
+SSRF protection is **on by default** — no feature flag needed. The `no-ssrf-protection` feature is an opt-in escape hatch. When compiled with it (e.g., for local development or testing), all IP blocklist checks are skipped.
+
+The blocklist logic is gated with `#[cfg(not(feature = "no-ssrf-protection"))]`:
```
Comment (pinodeca) [inline docs/spec-ssrf-protection.md]: This gate is used way too much. Instead of having it gate a lot of code, it should just gate the contents of the blocklist. When the feature is on, the blocklist should be empty.

File: src/activities/execute_http.rs [hunk 2]
```diff
@@ -8,19 +11,57 @@ use crate::types::HttpConfig;
...
+fn build_client(timeout: Duration) -> Result<reqwest::Client, String> {
+    let builder = reqwest::Client::builder().timeout(timeout);
+
+    #[cfg(not(feature = "no-ssrf-protection"))]
+    let builder = {
+        use crate::ssrf::{SsrfSafeResolver, SystemResolver};
+        use std::sync::Arc;
+        let resolver = SsrfSafeResolver::wrapping(Arc::new(SystemResolver));
+        builder.dns_resolver(Arc::new(resolver))
+    };
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply src/activities/execute_http.rs:30]: Fixed. Added `.redirect(reqwest::redirect::Policy::none())` to the client builder to prevent redirect-based SSRF bypasses. A 302 to an IP literal would skip the DNS resolver, so disabling redirects entirely is the safest approach.

File: docs/spec-ssrf-protection.md [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+## 2. Threat Model
+
+### Attack Scenarios
+
+**Cloud metadata exfiltration:**
+```sql
+SELECT df.start(
+    df.http('GET', 'http://169.254.169.254/latest/meta-data/iam/security-credentials/'),
+    'steal-creds'
+);
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply docs/spec-ssrf-protection.md:39]: Fixed. All examples in the spec now use `df.http(url, method)` — URL first, method second — matching the actual function signature in `src/dsl.rs`.

File: docs/spec-ssrf-protection.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+### 7.2 E2E Tests
+
+Create a test that verifies blocked requests fail with the expected error message:
+
+```sql
+-- Test: SSRF protection blocks link-local addresses
+SELECT df.start(
+    df.http('GET', 'http://169.254.169.254/latest/meta-data/'),
+    'test-ssrf-blocked'
+);
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply docs/spec-ssrf-protection.md]: Fixed — same change applied here.

File: docs/spec-ssrf-protection.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+    }
+}
+
+fn check_blocked_ipv6(ip: Ipv6Addr) -> Option<&'static str> {
+    if ip.is_loopback() {
+        return Some("loopback (::1)");
+    }
+    let segments = ip.segments();
+    if segments[0] == 0xfe80 {
+        return Some("link-local (fe80::/10)");
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply docs/spec-ssrf-protection.md]: Fixed. Updated the spec code sample to use `segments[0] & 0xffc0 == 0xfe80` (proper `/10` mask), matching the actual implementation in `src/ssrf.rs`.

File: src/activities/execute_http.rs [hunk 6]
```diff
@@ -50,6 +91,18 @@ pub async fn execute(ctx: ActivityContext, config_json: String) -> Result<String
...
+        let err_string = e.to_string();
+
+        // Detect SSRF IP-blocklist rejections from the resolver and emit
+        // a structured audit log (mirrors the scheme-block log above).
+        if err_string.contains("Blocked:") && err_string.contains("restricted") {
+            ctx.trace_info(format!(
+                "HTTP BLOCKED (ip) url={} submitted_by={audit_user} login_role={audit_login}",
+                config.url
+            ));
+            return err_string;
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply src/activities/execute_http.rs:109]: Fixed. Added `is_ssrf_block_error()` helper in `ssrf.rs` that owns the marker constants (`SSRF_BLOCK_MARKER` and `SSRF_RESTRICTED_MARKER`) right next to the resolver that emits them. `execute_http.rs` now calls `crate::ssrf::is_ssrf_block_error(&err_string)` instead of inline string matching.

File: docs/spec-ssrf-protection.md (hunk 1 above)
Comment (pinodeca) [reply docs/spec-ssrf-protection.md]: Fixed. Restructured `ssrf.rs` so that all functions, the resolver, and the system resolver exist unconditionally — no `#[cfg]` on any public API or type. The `no-ssrf-protection` feature now only gates the **contents** of `check_blocked_ipv4()` and `check_blocked_ipv6()`: when the feature is on, they return `None` (empty blocklist); when off, they return the actual blocked ranges. This also removed the `#[cfg]` from `validate_url_host`, `SsrfSafeResolver`, `SystemResolver`, and `build_client` in `execute_http.rs`.
